coldfront/core/department/utils/ldap.py

```diff
@@ -0,0 +1,115 @@
+        return conn.entries[0]
+    return None
+
+
+def get_L4_code_from_department_code(code):
+    """Return the L4 code for a department from a L4+ code.
+
+    Parameters:
+        code (str): L4+ department code
+    Returns:
+        (str) Normalized L4 department code
+    """
+    conn = Connection(LDAP_URL, auto_bind=True)
+    conn.search(DEPARTMENT_OU,
+                f'(&(objectClass=organizationalUnit)(ou={code}))',
+                attributes=['berkeleyEduOrgUnitHierarchyString'])
+    hierarchy = conn.entries[0].berkeleyEduOrgUnitHierarchyString.value
+    return hierarchy.split('-')[3]
+
+
+def get_department_name_from_code(code):
+    """Return the name of a department from a code.
+
+    Parameters:
+       code (str): department code
+    Returns:
+        (str) Name of department
+    """
+    conn = Connection(LDAP_URL, auto_bind=True)
+    conn.search(DEPARTMENT_OU,
+                f'(&(objectClass=organizationalUnit)(ou={code}))',
+                attributes=['description'])
+    return conn.entries[0].description.value
+
+
+def fetch_and_set_user_departments(user, userprofile, dry_run=False):
+    """Fetch a user's departments from LDAP and sets them in the
+    database.
+
+    Parameters:
+        user (User): Django User object
+        userprofile (UserProfile): Coldfront UserProfile object
+        dry_run (bool): If True, don't actually make changes
+                        (default: False)
+    Returns:
+        None
+    """
+    if dry_run:
+        logger.info = print
+        department = Department()
+        department.pk = '{placeholder_pk}'
+        userdepartment = UserDepartment()
+        userdepartment.pk = '{placeholder_pk}'
+    user_entry = ldap_search_user(user.email, user.first_name, user.last_name)
+    ldap_departments = user_entry.departmentNumber.values if user_entry else []
+    for department_code in ldap_departments:
+        department_code = get_L4_code_from_department_code(department_code)
+        # If a Department doesn't exist, create it.
+        name = None if Department.objects.filter(code=department_code).exists() \
+            else get_department_name_from_code(department_code)
+        if dry_run:
+            created = name is not None
+            department.code = department_code
+            department.name = department_code
+        else:
+            department, created = Department.objects.get_or_create(
+                                                    code=department_code,
+                                                    defaults={'name': name})
+        if created:
+            logger.info(f'Created Department {department.pk}, '
+                        f'{department.name} ({department.code})')
+
+        # Create a UserDepartment association, updating is_authoritative if
+        # needed.
+        if not dry_run:
+            userdepartment, created = UserDepartment.objects.update_or_create(
```

Comment:
Log that it was created or updated.

Reply:
I think we might have agreed on not logging creating UserDepartment associations. If we do, would I also have to log everywhere else were create UserDepartment associations? I'd also have to check if the association existed previously to see if it's been created.

Reply:
Okay, yes, we did agree on that--never mind.